python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -0,0 +1,311 @@
+    global_counter = 0
+
+    for logical_batch in logical_batch_iterator:
+        shuffle_buffer.append(logical_batch)
+        shuffle_buffer_size += logical_batch.num_rows
+
+        while shuffle_buffer_size >= shuffle_buffer_min_size:
+            output_batch = shuffle_buffer.pop(
+                random.randint(0, len(shuffle_buffer) - 1)
+            )
+            output_batch.batch_idx = global_counter
+            yield output_batch
+            shuffle_buffer_size -= output_batch.num_rows
+            global_counter += 1
+
+    # Yield any leftover.
+    while len(shuffle_buffer) > 0:
+        output_batch = shuffle_buffer.pop(random.randint(0, len(shuffle_buffer) - 1))
+        output_batch.batch_idx = global_counter
+        yield output_batch
+        global_counter += 1
+
+
+def prefetch_batches_locally(
+    logical_batch_iter: Iterator[LogicalBatch],
+    prefetcher: BlockPrefetcher,
+    num_batches_to_prefetch: int,
+) -> Iterator[LogicalBatch]:
+    """Given an iterator of logical batches, returns an iterator over the same logical
+    batches, while prefetching `num_batches_to_prefetch` batches in advance.
+
+    Args:
+        logical_batch_iter: An iterator over logical batches.
+        prefetcher: The prefetcher to use.
+        num_batches_to_prefetch: The number of batches to prefetch ahead of the
+            current batch during the scan.
+    """
+
+    def get_next_batches() -> Iterator[List[LogicalBatch]]:
+        """Return lists of logical batches corresponding to `num_batches_to_prefetch`"""
+        next_batches = []
+        while True:
+            try:
+                next_batches.append(next(logical_batch_iter))
+                if len(next_batches) == num_batches_to_prefetch:
+                    yield next_batches
+                    next_batches = []
+            except StopIteration:
+                break
+
+        if len(next_batches) > 0:
+            yield next_batches
+
+    # Fetch the initial set of batches.
+    batch_iterator = get_next_batches()
+    try:
+        batches = next(batch_iterator)
+    except StopIteration:
+        return
+
+    block_refs = [block_ref for batch in batches for block_ref in batch.block_refs]
+    prefetcher.prefetch_blocks(block_refs)
+
+    for next_batches in batch_iterator:
+        # Prefetch the next batches.
+        block_refs = [
+            block_ref for batch in next_batches for block_ref in batch.block_refs
+        ]
+        prefetcher.prefetch_blocks(block_refs)
+
+        for batch in batches:
+            yield batch
+
+        batches = next_batches
+
+    # Yield the final set of batches.
+    for batch in batches:
+        yield batch
+
+
+def resolve_logical_batch(
+    logical_batch_iter: Iterator[LogicalBatch],
+) -> Iterator[LogicalBatch]:
+    """Resolves the block references for each logical batch."""
+    for logical_batch in logical_batch_iter:
+        logical_batch.resolve()
+        yield logical_batch
```

Comment:
Ah I guess there's several stages of resolution:
1. Logical batch with unfetched blocks and unsliced blocks.
2. Partially resolved batch with fetched blocks and unsliced blocks.
3. Fully resolved batch with fetched and sliced blocks.

It might be cleaner to have these as three separate types: `LogicalBatch`, `ResolvedBatch`, and `Batch`.
```python
class LogicalBatch:
    batch_idx: int
    block_refs: List[ObjectRef[Block]]
    starting_block_idx: int
    ending_block_idx: Optional[int]
    num_rows: int


class ResolvedBatch:
    blocks: List[Block]
    logical_batch: LogicalBatch


class Batch:
    batch: Any
    logical_batch: LogicalBatch
```

This is a bit of a nit, so feel free to ignore.

Reply:
Removed `LogicalBatch` entirely